rpc/src/client/subscription.rs

```diff
@@ -0,0 +1,591 @@
+    /// the event's.
+    pub async fn publish(&mut self, ev: Event) {
+        let subs_for_query = match self.subscriptions.get_mut(&ev.query) {
+            Some(s) => s,
+            None => return,
+        };
+        let mut disconnected = Vec::<SubscriptionId>::new();
+        for (id, event_tx) in subs_for_query {
+            if event_tx.send(Ok(ev.clone())).await.is_err() {
+                disconnected.push(id.clone());
+            }
+        }
+        let subs_for_query = self.subscriptions.get_mut(&ev.query).unwrap();
+        for id in disconnected {
+            subs_for_query.remove(&id);
+        }
+    }
+
+    /// Immediately add a new subscription to the router without waiting for
+    /// confirmation.
+    pub fn add(&mut self, id: &SubscriptionId, query: String, event_tx: ChannelTx<Result<Event>>) {
+        let subs_for_query = match self.subscriptions.get_mut(&query) {
+            Some(s) => s,
+            None => {
+                self.subscriptions.insert(query.clone(), HashMap::new());
+                self.subscriptions.get_mut(&query).unwrap()
+            }
+        };
+        subs_for_query.insert(id.clone(), event_tx);
+    }
+
+    /// Keep track of a pending subscription, which can either be confirmed or
+    /// cancelled.
+    ///
+    /// `req_id` must be a unique identifier for this particular pending
+    /// subscription request operation, where `subs_id` must be the unique ID
+    /// of the subscription we eventually want added.
+    pub fn pending_add(
+        &mut self,
+        req_id: &str,
+        subs_id: &SubscriptionId,
+        query: String,
+        event_tx: ChannelTx<Result<Event>>,
+        result_tx: ChannelTx<Result<()>>,
+    ) {
+        self.pending_subscribe.insert(
+            req_id.to_string(),
+            PendingSubscribe {
+                id: subs_id.clone(),
+                query,
+                event_tx,
+                result_tx,
+            },
+        );
+    }
+
+    /// Attempts to confirm the pending subscription request with the given ID.
+    ///
+    /// Returns an error if it fails to respond to the original caller to
+    /// indicate success.
+    pub async fn confirm_add(&mut self, req_id: &str) -> Result<()> {
+        match self.pending_subscribe.remove(req_id) {
+            Some(mut pending_subscribe) => {
+                self.add(
+                    &pending_subscribe.id,
+                    pending_subscribe.query.clone(),
+                    pending_subscribe.event_tx,
+                );
+                Ok(pending_subscribe.result_tx.send(Ok(())).await?)
+            }
+            None => Ok(()),
+        }
+    }
+
+    /// Attempts to cancel the pending subscription with the given ID, sending
+    /// the specified error to the original creator of the attempted
+    /// subscription.
+    pub async fn cancel_add(&mut self, req_id: &str, err: impl Into<Error>) -> Result<()> {
```

Comment:
This is due to the possibility of concurrency issues when managing multiple subscriptions.

You could create one subscription and start receiving events for it, and then create a second subscription whose success/failure response from the RPC may only come back after an event relating to the first subscription.

Since events and subscription responses could be interspersed, and because subscriptions could possibly fail (implying the response from the RPC is important and we can't discard it), we need to facilitate a two-phase subscribe and unsubscribe process.

I'll see if I can add this into the ADR in a separate PR 👍